functions/validation/validation.js

```diff
@@ -0,0 +1,237 @@
+    errors.push({
+      message: ERRORS.INVALID_DATE,
+      offendingRelease: release,
+    });
+  } else if (new Date(release.releaseDate) <=
+    new Date(release.codeFreezeDate)) {
+    errors.push({
+      message: ERRORS.CODEFREEZE_AFTER_RELEASE,
+      offendingRelease: release,
+    });
+  }
+
+  return errors;
+}
+
+/**
+ * Check if all release names are unique. This function is not responsible for
+ * checking if these releases already exist in the database, only for checking
+ * that the release names are unique within the set of releases to be scheduled.
+ *
+ * This function assumes that the release names have already been validated.
+ *
+ * @param {Object} releases - The releases to validate.
+ * @return {Array} An array of error messages
+ */
+function validateUniqueReleaseNames(releases) {
+  const errors = [];
+  const releaseNames = releases.map((release) => release.releaseName);
+  const uniqueReleaseNames = new Set(releaseNames);
+
+  if (uniqueReleaseNames.size !== releaseNames.length) {
+    errors.push({
+      message: ERRORS.DUPLICATE_RELEASE_NAMES,
+    });
+  }
+
+  return errors;
+}
+
+/**
+ * Validates that a release object is in a valid form.
+ *
+ * @param {Object} release - The release to validate.
+ * @return {Array} errors - A list of errors from the validation of the release.
+ */
+function validateRelease(release) {
+  const releaseNameErrors = validateReleaseName(release);
+  const operatorErrors = validateReleaseOperator(release);
+  const dateErrors = validateReleaseDates(release);
+
+  return [...releaseNameErrors, ...operatorErrors, ...dateErrors];
+}
+
+/** Validation checks for a set of upcoming releases. This function is only
+   * intended to be used to validate releases that are to be scheduled, and not
+   * already existing releases.
+   *
+   * @param {Object} newReleases - A set of upcoming releases to be validated
+   * @return {Object} list of errors from the validation of the releases.
+   */
+function validateNewReleases(newReleases) {
+  const errors = [];
+
+  if (newReleases) {
+    for (const release of newReleases) {
+      const releaseErrors = validateRelease(release);
+      errors.push(...releaseErrors);
+    }
+
+    if (errors.length === 0) {
+      const uniqueReleaseNameErrors = validateUniqueReleaseNames(newReleases);
+      errors.push(...uniqueReleaseNameErrors);
+    }
+  } else {
+    errors.push({
+      message: ERRORS.NO_RELEASES,
+    });
+  }
+
+  return errors;
+}
+
+/**
+ * Validates new releases before they are stored in Firestore.
+ * This is primarily to ensure the date format is held, and not to ensure
+ * that any logical errors don't exist in the code.
+ *
+ * @param {Array} newReleases - New releases to validate
+ */
+function validateNewReleaseStructure(newReleases) {
+  if (!Array.isArray(newReleases)) {
+    throw new Error("New releases should be an array");
+  }
+
+  // Validate each release object
+  newReleases.forEach((release) => {
```

Comment:
A good functional behavior to pick up on is that anytime you find yourself mapping or looping through a collection- if the underlying action ends up being longer than like 3 lines, or if the behavior could be desirable in isolation- you should probably abstract the behavior into its own function.

For example:
```js
function validateNewReleaseStructure(newReleases) {
  if (!Array.isArray(newReleases)) {
    throw new Error("New releases should be an array");
  }

  newReleases.forEach(validateReleaseStructure)
}

function validateReleaseStructure(release) {
    if (typeof release != "object") {
      throw new Error("Each release should be an object");
    }

    if (!Object.prototype.hasOwnProperty.call(release, "releaseName") ||
      typeof release.releaseName !== "string") {
      throw new Error("Each release should have a string property"+
        " 'releaseName'");
    }

    if (!Object.prototype.hasOwnProperty.call(release, "releaseOperator") ||
      typeof release.releaseOperator !== "string") {
      throw new Error("Each release should have a string property" +
        " 'releaseName'");
    }

    if (!Object.prototype.hasOwnProperty.call(release, "codeFreezeDate") ||
      !(release.codeFreezeDate instanceof Timestamp)) {
      throw new Error("Each release should have a Firestore Timestamp"+
        " property 'codeFreezeDate'");
    }

    if (!Object.prototype.hasOwnProperty.call(release, "releaseDate") ||
      !(release.releaseDate instanceof Timestamp)) {
      throw new Error("Each release should have a Firestore Timestamp"+
        " property 'releaseDate'");
    }
}
```

Reply:
Resolved in https://github.com/firebase/firebase-release-dashboard/pull/8/commits/15c9d1d1cd01b3c812e1ca32042cbe083c0a0ce8